pkg/auth/providers/migration/activedirectory_migration.go

```diff
@@ -0,0 +1,243 @@
+package migration
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	crtbsByPrincipalAndUserIndex = "auth.management.cattle.io/crtbByPrincipalAndUser"
+	prtbsByPrincipalAndUserIndex = "auth.management.cattle.io/prtbByPrincipalAndUser"
+	tokenByUserRefKey            = "auth.management.cattle.io/token-by-user-ref"
+)
+
+type adMigration struct {
+	users        v3.UserInterface
+	userLister   v3.UserLister
+	tokenLister  v3.TokenLister
+	tokens       v3.TokenInterface
+	prtbs        v3.ProjectRoleTemplateBindingInterface
+	crtbs        v3.ClusterRoleTemplateBindingInterface
+	crtbIndexer  cache.Indexer
+	prtbIndexer  cache.Indexer
+	tokenIndexer cache.Indexer
+	authConfigs  v3.AuthConfigInterface
+	secrets      corev1.SecretInterface
+	certs        string
+	caPool       *x509.CertPool
+}
+
+// MigrateActiveDirectoryDNToGUID will go through all Rancher users and check to see if the principalID
+// is an LDAP distinguished name, which was the way we used to map Rancher users to their LDAP entries.
+// If a principalID is based on DN, this will update the user's principalID along with the tokens,
+// CRTBs, and PRTBs to use a principalID that is the objectGUID for the user.
+func MigrateActiveDirectoryDNToGUID(ctx context.Context, management *config.ManagementContext) {
+	m := adMigration{
```

Comment:
Nit: for visual clarity, I'd group field initializers by resource - tokens with tokenLister and so on.